chain/stmgr/read.go

```diff
@@ -10,146 +10,54 @@ import (
 	cbor "github.com/ipfs/go-ipld-cbor"
 
 	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/network"
 	"github.com/filecoin-project/lotus/chain/state"
 	"github.com/filecoin-project/lotus/chain/types"
 	"github.com/filecoin-project/specs-actors/actors/builtin/miner"
 	"github.com/filecoin-project/specs-actors/actors/util/adt"
 )
 
-type StateTreeCB func(state *state.StateTree) error
-
-func (sm *StateManager) WithParentStateTsk(tsk types.TipSetKey, cb StateTreeCB) error {
+func (sm *StateManager) ParentStateTsk(tsk types.TipSetKey) (*state.StateTree, error) {
 	ts, err := sm.cs.GetTipSetFromKey(tsk)
 	if err != nil {
-		return xerrors.Errorf("loading tipset %s: %w", tsk, err)
-	}
-
-	cst := cbor.NewCborStore(sm.cs.Blockstore())
-	state, err := state.LoadStateTree(cst, sm.parentState(ts))
-	if err != nil {
-		return xerrors.Errorf("load state tree: %w", err)
+		return nil, xerrors.Errorf("loading tipset %s: %w", tsk, err)
 	}
-
-	return cb(state)
+	return sm.ParentState(ts, cb)
 }
 
-func (sm *StateManager) WithParentState(ts *types.TipSet, cb StateTreeCB) error {
+func (sm *StateManager) ParentState(ts *types.TipSet) (*state.StateTree, error) {
 	cst := cbor.NewCborStore(sm.cs.Blockstore())
-	state, err := state.LoadStateTree(cst, sm.parentState(ts))
+	version := sm.GetNtwkVersion(context.TODO(), ts.Height()-1)
+	state, err := state.LoadStateTree(cst, sm.parentState(ts), version)
 	if err != nil {
-		return xerrors.Errorf("load state tree: %w", err)
+		return nil, xerrors.Errorf("load state tree: %w", err)
 	}
 
-	return cb(state)
+	return state, nil
 }
 
-func (sm *StateManager) WithStateTree(st cid.Cid, cb StateTreeCB) error {
```

Comment:
The callback-based API turned out to be more trouble than it was worth with the new abstractions. We can technically continue to support it, but then we'd basically be supporting two separate actor state abstractions.